{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1282a255-fa75-468f-8510-c9eeb5ea7814",
   "metadata": {},
   "source": [
    "# RNAseq Workshop with `bioconductor` and `DESeq2`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9786e33b-d39a-48e6-b5c0-aa82685d7e7c",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1194a828-4eb5-4f83-9480-94af07619975",
   "metadata": {},
   "source": [
    "### Instructor: **Brendan Jeffrey**\n",
    "\n",
    "### Email: brendan.jeffrey@nih.gov\n",
    "\n",
    "### Date: Jan 1, 2023"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07379752-d9d6-4886-b367-5d339ea61076",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae25fa2d-0aba-4f79-bf3f-1aa8df327203",
   "metadata": {},
   "source": [
    "### DESeq2 script execution instructions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acbf9c44-86f6-4215-a63f-e9bd435bb12f",
   "metadata": {},
   "source": [
    "Once R and Rstudio are installed, navigate to the code folder in the RNAseq_workshop_031819 folder and open the RNAseq_DESeq2_workshop.R script in Rstudio.\n",
    "\n",
    "1) Place cursor on line 8 and execute line by pressing command-enter\n",
    "1) Execute each individual package install, lines 13 through 25, using command-enter\n",
    "    1) For each package installation you may be prompted with “Update all/some/none? [a/s/n]:” - answer with a for all\n",
    "    1) May also be prompted with “Do you want to install from sources the packages which need compilation? (Yes/no/cancel)”. – answer no\n",
    "1) Once all packages installed you can execute each line of the R script, again using command-enter to execute the line that your cursor is currently on\n",
    "\n",
    "Please don’t hesitate to contact me if you are having any issues with R and Rstudio installation and R script execution – \n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa899ca5-a2a7-492f-92fa-cf1248a80071",
   "metadata": {},
   "source": [
    "### DESeq2\n",
    "\n",
    "DESeq2 manuscript: https://www.ncbi.nlm.nih.gov/pubmed/25516281"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca5026d4-b6b4-47ea-b95b-26b7dbe602ac",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96b20d80-5bc2-45fb-b490-b935c28a2827",
   "metadata": {},
   "source": [
    "## R and Rstudio installation\n",
    "\n",
    "The BCBB team has made R and Rstudio software available on the Self Service tool (under ‘Software’ tab) for Macs or the Software Center tool for windows. You should be able to install these tools without IT."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c8fb6ad-4978-4a3f-afd5-e6db70c3eb10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# package installation\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE)) {\n",
    "  install.packages(\"BiocManager\")\n",
    "}\n",
    "\n",
    "# # install packages with Bioconductor package manager\n",
    "# BiocManager::install(\"airway\", version = \"3.8\")\n",
    "# BiocManager::install(\"Rsamtools\", version = \"3.8\")\n",
    "# BiocManager::install(\"GenomicFeatures\", version = \"3.8\")\n",
    "# BiocManager::install(\"DESeq2\", version = \"3.8\")\n",
    "# BiocManager::install(\"vsn\", version = \"3.8\")\n",
    "# BiocManager::install(\"hexbin\", version = \"3.8\")\n",
    "# BiocManager::install(\"pheatmap\", version = \"3.8\")\n",
    "# BiocManager::install(\"apeglm\", version = \"3.8\")\n",
    "# BiocManager::install(\"org.Hs.eg.db\", version = \"3.8\")\n",
    "# BiocManager::install(\"ReportingTools\", version = \"3.8\")\n",
    "# BiocManager::install(\"Gviz\", version = \"3.8\")\n",
    "# BiocManager::install(\"ggbeeswarm\", version = \"3.8\")\n",
    "# BiocManager::install(\"EnrichmentBrowser\", version = \"3.8\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ade11b80-44e5-48c2-bf31-5c252ef70850",
   "metadata": {},
   "source": [
    "## Part 1 generating the count matrix, summarized experiment, sample information2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e928c9ec-11b8-44c6-95df-1d6161c4a545",
   "metadata": {},
   "source": [
    "Generating summarizedExperiment object METHOD 1, using summarizeOverlaps\n",
    "\n",
    "- Airway study file and sample information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91406677-4909-4559-b183-563f8533a3a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load airway package - see https://bioconductor.org/packages/3.8/data/experiment/html/airway.html\n",
    "# \"airway\" manuscript - https://www.ncbi.nlm.nih.gov/pubmed/24926665\n",
    "library(\"airway\")\n",
    "\n",
    "# get information about a package\n",
    "browseVignettes(\"airway\")\n",
    "\n",
    "# where is this data located, and list the files in that directory\n",
    "indir <- system.file(\"extdata\", package=\"airway\", mustWork=TRUE)\n",
    "indir\n",
    "list.files(indir)\n",
    "\n",
    "# file with detailed information about each of our samples - sample_table\n",
    "csvfile <- file.path(indir, \"sample_table.csv\")\n",
    "csvfile\n",
    "sampleTable <- read.csv(csvfile, row.names = 1)\n",
    "\n",
    "# print out contents\n",
    "sampleTable\n",
    "\n",
    "filenames <- file.path(indir, paste0(sampleTable$Run, \"_subset.bam\"))\n",
    "sampleTable$Run\n",
    "\n",
    "file.exists(filenames)\n",
    "# what else could you have used for generating the bam files?\n",
    "# filenames <- file.path(indir, paste0(row.names(sampleTable), \"_subset.bam\"))\n",
    "\n",
    "library(\"Rsamtools\")\n",
    "# indicate that these are bam files usiing BamFileList function from Rsamtools package\n",
    "# only process 2 million reads at a time\n",
    "bamfiles <- BamFileList(filenames, yieldSize=2000000)\n",
    "bamfiles\n",
    "\n",
    "# NOTE: make sure that the chromosome names of the genomic features in the annotation \n",
    "# you use are consistent with the chromosome names of the reference used for read alignment\n",
    "# check chromosome names (here called seqnames)\n",
    "seqinfo(bamfiles[1])\n",
    "\n",
    "# Define gene models - read from GTF file http://www.ensembl.org/info/website/upload/gff.html\n",
    "library(\"GenomicFeatures\")\n",
    "\n",
    "# genomic coord and gene file for airway study (GTF)\n",
    "gtffile <- file.path(indir,\"Homo_sapiens.GRCh37.75_subset.gtf\")\n",
    "\n",
    "# load from this gtf file and indicate none of our sequences are circular\n",
    "txdb <- makeTxDbFromGFF(gtffile, format = \"gtf\", circ_seqs = character())\n",
    "?makeTxDbFromGFF\n",
    "\n",
    "# make a GRangesList of all exons grouped by gene\n",
    "# get more information on a function\n",
    "?exonsBy\n",
    "ebg <- exonsBy(txdb, by=\"gene\")\n",
    "ebg\n",
    "\n",
    "# Read counting step using summarizeOverlaps function from GenomicAlignments package\n",
    "# NOTE: can use multiple cores to speed up if desired using BiocParallel package https://bioconductor.org/packages/3.8/BiocParallel\n",
    "\n",
    "# load GenomicAlignments package\n",
    "library(\"GenomicAlignments\")\n",
    "\n",
    "# create the SummarizedExperiment object \n",
    "se <- summarizeOverlaps(features=ebg, reads=bamfiles,\n",
    "                        mode=\"Union\",\n",
    "                        singleEnd=FALSE,\n",
    "                        ignore.strand=TRUE,\n",
    "                        fragments=TRUE )\n",
    "\n",
    "# explore options of summarizeOverlaps\n",
    "?summarizeOverlaps\n",
    "\n",
    "# other tools that could be used:\n",
    "# summarizeOverlaps, GenomicAlignments package, 'DESeqDataSet' function\n",
    "# featureCounts, Rsubread package, 'DESeqDataSetFromMatrix' function\n",
    "# tximport, tximport package, 'DESeqDataSetFromTximport' function\n",
    "# htseq-count, HTSeq Python package (not part of R), 'DESeqDataSetFromHTSeq' function\n",
    "\n",
    "# explore the SummarizedExperiment object\n",
    "browseVignettes(\"SummarizedExperiment\")\n",
    "se\n",
    "\n",
    "# look at assay - the count matrix\n",
    "assayNames(se)\n",
    "assay(se)\n",
    "\n",
    "# rowRanges slot of se, contains information on the genomomic ranges, ie each gene\n",
    "rowRanges(se)\n",
    "\n",
    "# colData slot of se (currently empty), contains information about the samples\n",
    "colData(se)\n",
    "\n",
    "# populate the colData slot with information from sampleTable\n",
    "sampleTable\n",
    "\n",
    "# Because we used a column of  sampleTable to produce the bamfiles vector, \n",
    "# we know the columns of se are in the same order as the rows of sampleTable\n",
    "colData(se) <- DataFrame(sampleTable)\n",
    "colData(se)\n",
    "\n",
    "#### generating DESeqDataSet object METHOD 2, pre-existing count matrix ####\n",
    "# read in counts matrix (HTSeq generated, STAR, others)\n",
    "counts <- read.table(\"../data/RNAseq_counts_matrix.txt\", header = TRUE, row.names=1)\n",
    "counts\n",
    "\n",
    "# read in sample data\n",
    "sample_table <- read.table(\"../data/RNAseq_experiment_design.csv\", header=TRUE, sep=',')\n",
    "sample_table\n",
    "\n",
    "# experimental design\n",
    "exp_design <- data.frame(row.names = sample_table$replicate, \n",
    "                         condition = sample_table$condition,\n",
    "                         sample_names = sample_table$sample_name)\n",
    "exp_design\n",
    "\n",
    "# generate the DESeqDataSet using the DESeqDataSetFromMatrix function in DESeq2\n",
    "# DESeqDataSet is a custom class within DESeq2, built on top of the generalize summarizedExperiment class\n",
    "library(\"DESeq2\")\n",
    "?DESeqDataSetFromMatrix\n",
    "dds <- DESeqDataSetFromMatrix(countData = counts, colData = exp_design, design = ~ condition)\n",
    "\n",
    "# explore the DESeqDataSet, and compare to summarizedExperiment above\n",
    "dds\n",
    "se\n",
    "colData(dds)\n",
    "colData(se)\n",
    "\n",
    "rowRanges(dds)\n",
    "rowRanges(se)\n",
    "\n",
    "#### END generating summarized experiment, DESeqDataSet ####"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfeeffed-18df-43db-a4ef-916af39f5d08",
   "metadata": {},
   "source": [
    "### BRANCH point\n",
    "\n",
    "After generating the summarizedExperiment (DESeqDataSet) can use a variety of Bioconductor packages for differential gene expression. These include:\n",
    "\n",
    "    - edgeR https://bioconductor.org/packages/3.8/edgeR\n",
    "    - limma https://bioconductor.org/packages/3.8/limma\n",
    "    - EBSeq https://bioconductor.org/packages/3.8/EBSeq - isoform differential expression"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fdeeaf6-ccfc-4c6b-ace9-1504a09e1660",
   "metadata": {},
   "source": [
    "## Part 2 using DESeq2 for data exploration "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1ad3743-8929-4a47-bfbd-91e6e7eb0203",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### starting from a prepared SummarizedExperiment ####\n",
    "# similar to SummarizedExperiment above, except using all genes\n",
    "# clear environment\n",
    "\n",
    "# load airway summarized experiment data\n",
    "data(\"airway\")\n",
    "airway\n",
    "\n",
    "# rename SummarizedExperiment object\n",
    "se <- airway\n",
    "\n",
    "# explore\n",
    "rowRanges(se)\n",
    "colData(se)\n",
    "\n",
    "# we want to specify that untrt (untreated) is the reference level for dex (dexamethasone)\n",
    "se$dex\n",
    "se$dex <- relevel(se$dex, \"untrt\")\n",
    "se$dex\n",
    "\n",
    "# check the millions of fragments that uniquely aligned to genes, one decimal\n",
    "round( colSums(assay(se)) / 1e6, 1 )\n",
    "\n",
    "# construct the DESeqDataSet object from the SummarizedExperiment object\n",
    "# DESeqDataSet is a custom class within DESeq2, built on top of the generalized summarizedExperiment object\n",
    "dds <- DESeqDataSet(se, design = ~ cell + dex)\n",
    "\n",
    "# The simplest design formula for differential expression would be ~ condition, \n",
    "# where  condition is a column in colData(dds) that specifies which of two (or more groups) \n",
    "# the samples belong to. For the airway experiment, we will specify ~ cell + dex \n",
    "# meaning that we want to test for the effect of dexamethasone (dex) \n",
    "# controlling for the effect of different cell line (cell)\n",
    "# also used when there are batch effects present in data\n",
    "\n",
    "# pre-filter the counts data to remove those genes with zero counts\n",
    "nrow(dds)\n",
    "\n",
    "# slice of the dds values [row, column]\n",
    "dds <- dds[ rowSums(counts(dds)) > 1, ]\n",
    "nrow(dds)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "067d3407-d6f0-4f62-b4a9-5f6e773fdf8f",
   "metadata": {},
   "source": [
    "### Exploratory analysis and visualization\n",
    "\n",
    "- Variance stabilization transformation and rlog\n",
    "- Common statistical methods for exploratory analysis, eg clustering and principal component analysis (PCA)\n",
    "- Work best with data that has the same range of variance at different ranges of means (homoskedastic)\n",
    "- For RNAseq data, expected variance grows with the mean\n",
    "- Performing PCA on raw counts, resulting plot depends mostly on genes with highest counts\n",
    "- Performing PCA on simple log transformed counts plus pseudo count of 1, the genes with low counts\n",
    "- Will contribute a lot of noise, due to taking logarithm of small counts inflates their variance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1753580-85e4-4228-9e90-a7089398ff0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Effects of transformations on the variance ##\n",
    "# vizualizing with simulated data\n",
    "library(\"vsn\")\n",
    "lambda <- 10^seq(from = -1, to = 2, length = 1000)\n",
    "cts <- matrix(rpois(1000*100, lambda), ncol = 100)\n",
    "\n",
    "# plot on non-transformed\n",
    "meanSdPlot(cts, ranks = FALSE)\n",
    "\n",
    "# plot on log transformed count + 1\n",
    "log.cts.one <- log2(cts + 1)\n",
    "meanSdPlot(log.cts.one, ranks = FALSE)\n",
    "\n",
    "# DESeq2 offers two transformation methods, VST and RLD\n",
    "# VST (variance stabilizing transformation) faster, for medium to large datasets\n",
    "# RLD (Regularized-logarithm transformation), better for small datasets, n < 30\n",
    "\n",
    "# VST - transform our data\n",
    "vsd <- vst(dds, blind = FALSE)\n",
    "head(assay(vsd), 3)\n",
    "meanSdPlot(assay(vsd), ranks = FALSE)\n",
    "\n",
    "# compare to raw read count\n",
    "meanSdPlot(assay(dds), ranks = FALSE)\n",
    "\n",
    "#### Sample distances ####\n",
    "# A useful first step in an RNA-Seq analysis is often to assess \n",
    "# overall similarity between samples: Which samples are similarto each other, \n",
    "# which are different? Does this fit to the expectation from the experiment’s design?\n",
    "# R function dist to calculate the Euclidian distance between samples\n",
    "# t used to transpose the results of assay(rld), samples are now rows\n",
    "sampleDists <- dist(t (assay(vsd)) )\n",
    "sampleDists\n",
    "\n",
    "# vizualize the distance matrix using pheatmap package\n",
    "library(\"pheatmap\")\n",
    "library(\"RColorBrewer\")\n",
    "\n",
    "# sample distances to matrix, base R as.matrix function\n",
    "sampleDistMatrix <- as.matrix(sampleDists)\n",
    "sampleDistMatrix\n",
    "\n",
    "# plot, , specify rownames and colnames for clarity\n",
    "sampleDistMatrix\n",
    "rownames(sampleDistMatrix) <- paste( vsd$dex, vsd$cell, sep = \" - \" )\n",
    "colnames(sampleDistMatrix) <- NULL\n",
    "sampleDistMatrix\n",
    "\n",
    "# default colors\n",
    "pheatmap(sampleDistMatrix,\n",
    "         clustering_distance_rows = sampleDists,\n",
    "         clustering_distance_cols = sampleDists)\n",
    "\n",
    "# setting colors using RcolorBrewer\n",
    "colors <- colorRampPalette( rev(brewer.pal(9, \"Blues\")) )(255)\n",
    "pheatmap(sampleDistMatrix,\n",
    "         clustering_distance_rows = sampleDists,\n",
    "         clustering_distance_cols = sampleDists,\n",
    "         col = colors)\n",
    "\n",
    "\n",
    "## visualize sample to sample distances using Principal component analysis, DESeq2 plotPCA function ##\n",
    "colData(vsd)\n",
    "plotPCA(vsd, intgroup = \"dex\")\n",
    "plotPCA(vsd, intgroup = c(\"dex\", \"cell\"))\n",
    "\n",
    "# quick aside, saving figures\n",
    "library(\"ggplot2\")\n",
    "\n",
    "plt <- plotPCA(vsd, intgroup = c(\"dex\", \"cell\"))\n",
    "plt\n",
    "ggsave(\"../results/default_PCA.pdf\", height=4, width=6.5)\n",
    "\n",
    "# more control over plotting with ggplot\n",
    "pcaData <- plotPCA(vsd, intgroup = c( \"dex\", \"cell\"), returnData = TRUE)\n",
    "pcaData\n",
    "percentVar <- round(100 * attr(pcaData, \"percentVar\"))\n",
    "percentVar\n",
    "\n",
    "# build plot\n",
    "gplt <- ggplot(pcaData, aes(x = PC1, y = PC2, color = dex, shape = cell)) +\n",
    "  geom_point(size =3) +\n",
    "  xlab(paste0(\"PC1: \", percentVar[1], \"% variance\")) +\n",
    "  ylab(paste0(\"PC2: \", percentVar[2], \"% variance\")) +\n",
    "  coord_fixed()\n",
    "gplt\n",
    "ggsave(\"../results/ggplot_PCA.pdf\", height=4, width=6.5)\n",
    "\n",
    "# ggplot cheat cheet - https://www.rstudio.com/wp-content/uploads/2015/03/ggplot2-cheatsheet.pdf\n",
    "\n",
    "\n",
    "## Heatmap of gene variance estimates ##\n",
    "# extract row indexes of those genes that have largest variance estimates\n",
    "topVarGenes <- head(order(rowVars(assay(vsd)), decreasing = TRUE), 20)\n",
    "topVarGenes\n",
    "\n",
    "# subset of the vsd values [row, column]\n",
    "mat <- assay(vsd)[ topVarGenes, ]\n",
    "mat\n",
    "\n",
    "# how does each genes transformed count deviate from the genes average across all samples\n",
    "mat <- mat - rowMeans(mat)\n",
    "mat\n",
    "anno <- as.data.frame(colData(vsd)[, c(\"cell\",\"dex\")])\n",
    "anno\n",
    "pheatmap(mat, annotation_col = anno)\n",
    "?pheatmap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93339c89-f2c7-42c2-83e4-5a3c0751d41f",
   "metadata": {},
   "source": [
    "### Part 3 using DESeq2 for differential expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "631c46ea-1771-4b0e-af4c-fbe9bd6ef129",
   "metadata": {},
   "outputs": [],
   "source": [
    "## DESeq2 differential expression with DESeq function ##\n",
    "dds <- DESeq(dds)\n",
    "?DESeq\n",
    "\n",
    "# rowData of dds now contains all fitted parameters\n",
    "head(rowData(dds))\n",
    "\n",
    "# extract results from dds DESeq2 results function\n",
    "res <- results(dds)\n",
    "res\n",
    "\n",
    "# generating results with more specific command, more than one treatment, forcing denominator\n",
    "res <- results(dds, contrast=c(\"dex\",\"trt\",\"untrt\"))\n",
    "\n",
    "# DESeq2 performs for each gene a hypothesis test to see whether evidence \n",
    "# is sufficient to decide against the null hypothesis that there is zero effect \n",
    "# of the treatment on the gene and that the observed difference between \n",
    "# treatment and control was merely caused by experimental variability\n",
    "\n",
    "# summary of the results\n",
    "summary(res)\n",
    "\n",
    "# DESeq2 uses the Benjamini-Hochberg (BH) adjustment for FDR\n",
    "# adjusting thresholds, false discovery rate threshold\n",
    "res.05 <- results(dds, alpha = 0.05)\n",
    "summary(res.05)\n",
    "\n",
    "# subset results, consider a fraction of 10% false positives acceptable FDR alpha=0.1\n",
    "resSig <- subset(res, padj < 0.1)\n",
    "summary(resSig)\n",
    "\n",
    "# extract up regulated genes\n",
    "resSigUp <- subset(resSig, log2FoldChange > 0)\n",
    "resSigUp\n",
    "\n",
    "# extract down regulated\n",
    "resSigDown <- subset(resSig, log2FoldChange < 0)\n",
    "resSigDown\n",
    "\n",
    "## plotting results ##\n",
    "# counts plot of individual genes\n",
    "library(\"ggbeeswarm\")\n",
    "topGene <- rownames(resSig)[which.min(resSig$padj)]\n",
    "topGene\n",
    "geneCounts <- plotCounts(dds, gene = topGene, intgroup = c(\"dex\",\"cell\"), returnData = TRUE)\n",
    "\n",
    "# generate plot\n",
    "countplt <- ggplot(geneCounts, aes(x = dex, y = count, color = cell, group = cell)) +\n",
    "  scale_y_log10() + \n",
    "  geom_point(size = 3) + \n",
    "  geom_line() + \n",
    "  ggtitle(topGene)\n",
    "countplt\n",
    "\n",
    "####### MA plots? M (log ratio) and A (mean average)  ####### \n",
    "# overview of the distribution of the estimated coefficients, or comparisons of interest, across all genes\n",
    "library(\"apeglm\")\n",
    "\n",
    "# show model coefficients\n",
    "resultsNames(dds)\n",
    "\n",
    "# moderate or shrink log2 fold changes\n",
    "resMA <- lfcShrink(dds, coef=\"dex_trt_vs_untrt\", type=\"apeglm\")\n",
    "plotMA(resMA, ylim = c(-5, 5))\n",
    "\n",
    "# compare with no statistical moderation to shrink noisy log2 fold change\n",
    "plotMA(res, ylim = c(-5, 5))\n",
    "\n",
    "################################\n",
    "### everyone loves a heatmap ###\n",
    "# heatmap of top 50 significant genes\n",
    "topSigGenes <- head(rownames(resSig[order(resSig$padj),]), 50)\n",
    "topSigGenes\n",
    "\n",
    "# extract VST transformed value for all genes for each sample\n",
    "assay(vsd)\n",
    "matSig <- assay(vsd)[topSigGenes, ]\n",
    "matSig\n",
    "\n",
    "# how does each genes transformed count deviate from the genes average across all samples\n",
    "matSig <- matSig - rowMeans(matSig)\n",
    "matSig\n",
    "\n",
    "# annotation from vsd, remember [row, column]\n",
    "anno <- as.data.frame(colData(vsd)[ , c(\"cell\",\"dex\")])\n",
    "anno\n",
    "\n",
    "# generate heatmap\n",
    "pheatmap(matSig, annotation_col = anno)\n",
    "\n",
    "#### Annotating and exporting results ####\n",
    "library(\"AnnotationDbi\")\n",
    "library(\"org.Hs.eg.db\")\n",
    "\n",
    "# annotation keys available\n",
    "columns(org.Hs.eg.db)\n",
    "\n",
    "rowRanges(dds)$symbol <- mapIds(org.Hs.eg.db,\n",
    "                            keys=row.names(dds),\n",
    "                            column=\"SYMBOL\",\n",
    "                            keytype=\"ENSEMBL\",\n",
    "                            multiVals=\"first\")\n",
    "\n",
    "\n",
    "\n",
    "# add gene symbol\n",
    "resSig$symbol <- mapIds(org.Hs.eg.db,\n",
    "                     keys=row.names(resSig),\n",
    "                     column=\"SYMBOL\",\n",
    "                     keytype=\"ENSEMBL\",\n",
    "                     multiVals=\"first\")\n",
    "\n",
    "# add entrez ID\n",
    "resSig$entrez <- mapIds(org.Hs.eg.db,\n",
    "                     keys=row.names(resSig),\n",
    "                     column=\"ENTREZID\",\n",
    "                     keytype=\"ENSEMBL\",\n",
    "                     multiVals=\"first\")\n",
    "\n",
    "# google search on entrez gene to check\n",
    "resSig\n",
    "\n",
    "# export csv\n",
    "resOrdered <- resSig[order(resSig$pvalue),]\n",
    "head(resOrdered)\n",
    "\n",
    "resOrderedDF <- as.data.frame(resOrdered)[1:100, ]\n",
    "write.csv(resOrderedDF, file = \"../results/results.csv\")\n",
    "\n",
    "## nice HTML reporting\n",
    "library(\"ReportingTools\")\n",
    "htmlRep <- HTMLReport(shortName=\"report\", title=\"My report\",\n",
    "                      reportDirectory=\"../results/report\")\n",
    "publish(resOrderedDF, htmlRep)\n",
    "url <- finish(htmlRep)\n",
    "browseURL(url)\n",
    "\n",
    "\n",
    "### additional plotting - plotting changes in genomic space ###\n",
    "resGR <- results(dds, format=\"GRanges\")\n",
    "resGR$log2FoldChange <- res$log2FoldChange\n",
    "resGR\n",
    "\n",
    "# add annotation again\n",
    "resGR$symbol <- mapIds(org.Hs.eg.db, names(resGR), \"SYMBOL\", \"ENSEMBL\")\n",
    "library(\"Gviz\")\n",
    "\n",
    "# 1 million base pairs up and downstream from gene with smallest pvalue\n",
    "topGene <- which.min(resGR$padj)\n",
    "\n",
    "window <- resGR[topGene] + 1e6\n",
    "strand(window) <- \"*\"\n",
    "resGRsub <- resGR[resGR %over% window]\n",
    "\n",
    "# if gene has no symbol or is duplicated\n",
    "naOrDup <- is.na(resGRsub$symbol) | duplicated(resGRsub$symbol)\n",
    "naOrDup\n",
    "\n",
    "# use the gene name if no symbol, else use symbol name\n",
    "resGRsub$group <- ifelse(naOrDup, names(resGRsub), resGRsub$symbol)\n",
    "resGRsub\n",
    "\n",
    "# create a vector specifying if the genes in this window had a low padj value\n",
    "status <- factor(ifelse(resGRsub$padj < 0.1 & !is.na(resGRsub$padj), \"sig\", \"notsig\"))\n",
    "status\n",
    "\n",
    "# plot using Gviz\n",
    "options(ucscChromosomeNames = FALSE)\n",
    "g <- GenomeAxisTrack()\n",
    "a <- AnnotationTrack(resGRsub, name = \"gene ranges\", feature = status)\n",
    "d <- DataTrack(resGRsub, data = \"log2FoldChange\", baseline = 0,\n",
    "               type = \"h\", name = \"log2 fold change\", strand = \"+\")\n",
    "plotTracks(list(g, d, a), groupAnnotation = \"group\", notsig = \"grey\", sig = \"hotpink\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
